lib/shared/sdk/usbboot/index.js

```diff
@@ -193,251 +133,318 @@ const usbIdToString = (id) => {
+   * @fulfil {Buffer} - blob
+   * @returns {Promise}
+   *
+   * @example
+   * const Bluebird = require('bluebird')
+   * const fs = Bluebird.promisifyAll(require('fs'))
+   *
+   * const blob = adapter.queryBlobFromCache('start.elf', (name) => {
+   *   return fs.readFileAsync(path.join('./blobs', name))
+   * })
+   */
+  queryBlobFromCache (name, fallback) {
+    if (this.blobCache[name]) {
+      return Bluebird.resolve(this.blobCache[name])
+    }
 
-      if (fileMessage.command === protocol.FILE_MESSAGE_COMMANDS.DONE) {
-        debug('Done')
-        return Bluebird.resolve()
+    return fallback(name).tap((buffer) => {
+      this.blobCache[name] = buffer
+    })
+  }
+
+  /**
+   * @summary Scan for usbboot capable USB devices
+   * @public
+   *
+   * @description
+   * You should at the very least pass a file named `bootcode.bin`.
+   *
+   * @param {Object} options - options
+   * @param {Object} options.files - files buffers
+   * @param {Function} callback - callback
+   * @returns {USBBootAdapter}
+   *
+   * @example
+   * adapter.scan({
+   *   files: {
+   *     'bootcode.bin': fs.readFileSync('./msd/bootcode.bin'),
+   *     'start.elf': fs.readFileSync('./msd/start.elf')
+   *   }
+   * }, (error, devices) => {
+   *   // ...
+   * })
+   */
+  scan (options = {}, callback) {
+    /* eslint-disable lodash/prefer-lodash-method */
+    usb.listDevices().filter(isUsbBootCapableUSBDevice).map((device) => {
+    /* eslint-enable lodash/prefer-lodash-method */
+
+      const idPair = _.join([
+        usbIdToString(device.deviceDescriptor.idVendor),
+        usbIdToString(device.deviceDescriptor.idProduct)
+      ], ':')
+
+      device.device = idPair
+      device.displayName = 'Initializing device'
+      device.description = 'Compute Module'
+      device.raw = idPair
+      device.size = null
+      device.mountpoints = []
+      device.protected = false
+      device.system = false
+      device.disabled = true
+      device.icon = 'loading'
+      device.adapter = this.id
+
+      // We need to open the device in order to access _configDescriptor
+      debug(`Opening device: ${device.displayName}`)
+      device.open()
+
+      // Ensures we don't wait forever if an issue occurs
+      device.timeout = USB_OPERATION_TIMEOUT_MS
+
+      // Handle 2837 where it can start with two interfaces, the first
+      // is mass storage the second is the vendor interface for programming
+      const addresses = {}
+      /* eslint-disable no-underscore-dangle */
+      if (device._configDescriptor.bNumInterfaces === USB_ENDPOINT_INTERFACES_SOC_BCM2835) {
+      /* eslint-enable no-underscore-dangle */
+        addresses.interface = 0
+        addresses.endpoint = 1
+      } else {
+        addresses.interface = 1
+        addresses.endpoint = 3
       }
 
+      const deviceInterface = device.interface(addresses.interface)
+      debug(`Claiming interface: ${addresses.interface}`)
+      deviceInterface.claim()
```

Comment:
Looks like we lost the `catch` block that is on `master`: https://github.com/resin-io/etcher/blob/master/lib/shared/sdk/usbboot/index.js#L404